crates/sui-types/src/effects/effects_aux_data.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::base_types::ObjectRef;
+use crate::crypto::default_hash;
+use crate::digests::{EffectsAuxDataDigest, TransactionDigest};
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone, Hash)]
+pub struct EffectsAuxData<T> {
+    tx_digest: TransactionDigest,
+    version: u64,
+    inner: T,
+}
+
+/// Separating storage type and in-memory type allows for customized serde.
+pub type EffectsAuxDataInStorage = EffectsAuxData<Vec<u8>>;
```

Comment:
Why do we need customized serde? Isn't `EffectsAuxData<Vec<EffectsAuxDataEntry>>` sufficiently evolvable? We can simply add arbitrary new enumerants to extend the type.

Reply:
For example, let's say a lot of the entries share some common object ID or addresses. Then you may want to have a lookup index table (in storage) that stores all the object IDs/addresses, and refer to them through index in each entry.